core/src/main/kotlin/org/evomaster/core/search/algorithms/MosaAlgorithm.kt

```diff
@@ -108,78 +140,147 @@ class MosaAlgorithm<T> : SearchAlgorithm<T>() where T : Individual {
+     *  Selection of the Targets"
+     */
+    private fun compare(x: Data, y: Data, notCovered: Set<Int>): Int {
+        var dominatesX = false
+        var dominatesY = false
+
+        for (index in 1..notCovered.size) {
+            if (x.ind.fitness.getHeuristic(index) > y.ind.fitness.getHeuristic(index))
+                dominatesX = true
+            if (y.ind.fitness.getHeuristic(index) > x.ind.fitness.getHeuristic(index))
+                dominatesY = true
+
+            // if the both do not dominates each other, we don't
+            // need to iterate over all the other targets
+            if (dominatesX && dominatesY)
+                return 0
+        }
+
+        if (dominatesX == dominatesY)
+            return 0
+
+        else if (dominatesX)
+            return -1
+
+        else (dominatesY)
+            return +1
+    }
+
+    private fun mosaPreferenceCriterion(notCovered: Set<Int>, list: List<Data>): HashSet<Data> {
+        var frontZero: HashSet<Data> = HashSet<Data>()
+
         notCovered.forEach { t ->
             var chosen = list[0]
             list.forEach { data ->
                 if (data.ind.fitness.getHeuristic(t) > chosen.ind.fitness.getHeuristic(t)) {
-                    //recall: maximization problem
+                    // recall: maximization problem
+                    chosen = data
+                } else if (data.ind.fitness.getHeuristic(t) == chosen.ind.fitness.getHeuristic(t)
+                            && data.ind.individual.seeGenes().size < chosen.ind.individual.seeGenes().size){
+                    // Secondary criterion based on tests lengths
                     chosen = data
                 }
             }
-            //MOSA preference criterion: the best for a target gets Rank 0
+            // MOSA preference criterion: the best for a target gets Rank 0
             chosen.rank = 0
+            frontZero.add(chosen)
         }
-    }
-
-    private fun dominates(d: Data, z: Data, notCovered: Set<Int>): Boolean {
-        return d.ind.fitness.subsumes(z.ind.fitness, notCovered)
+        return frontZero
     }
 
     private fun selection(): EvaluatedIndividual<T> {
 
-        //assumed sorted population, where first are bests
-        var min = population.size
+        // the population is not fully sorted
+        var min = randomness.nextInt(population.size)
 
-        (0 until config.tournamentSize).forEach {
+        (0 until config.tournamentSize-1).forEach {
             val sel = randomness.nextInt(population.size)
-            if (sel < min) {
+            if (population[sel].rank < population[min].rank) {
                 min = sel
+            } else if (population[sel].rank == population[min].rank){
+                if (population[sel].crowdingDistance < population[min].crowdingDistance)
+                    min = sel
             }
         }
 
-        return population[min].copy()
+        return (population[min].ind as EvaluatedIndividual<T>).copy()
```

Comment:
"as" is just a cast. but are you not adding "Data" to the population. Have you try to run EvoMaster with MOSA? u d need to use "--algorithm MOSA"

Reply:
The individual is added to the population later in the method "Search()" as follows:
    _var ind = selection()
    var eval = getMutatator().mutateAndSave(ind, archive)
    nextPop.add(Data(eval as EvaluatedIndividual))_

I have run indeed EvoMaster with the parameter "--algorithm MOSA" and it works on my laptop to work after these changes